lib/OpenQA/WebAPI/Controller/Admin/Needle.pm

```diff
@@ -74,69 +118,10 @@ sub ajax {
-            while (my $n = $needles->next) {
-                my $filename = $n->filename;
-                my $hash     = {
-                    id        => $n->id,
-                    directory => $n->directory->name,
-                    filename  => $filename,
-                };
-                my $needle_dir_path = $n->directory->path;
-                my $real_needle_dir_id;
-
-                if ($needle_dir_real_paths{$needle_dir_path}) {
-                    my $real_path = $needle_dir_real_paths{$needle_dir_path}->{real_path};
-                    $real_needle_dir_id = $needle_dir_real_paths{$needle_dir_path}->{real_path_id}
-                      if $real_path && $needle_dir_path ne $real_path;
-                }
-                else {
-                    my $real_path_id         = $n->directory->id;
-                    my $needle_dir_real_path = realpath($needle_dir_path);
-                    if ($needle_dir_real_path && $needle_dir_real_path ne $needle_dir_path) {
-                        my $real_needle_dir
-                          = $self->schema->resultset('NeedleDirs')->find({path => $needle_dir_real_path});
-                        if ($real_needle_dir) {
-                            $real_needle_dir_id = $real_needle_dir->id;
-                            $real_path_id       = $real_needle_dir_id;
-                        }
-                    }
-                    $needle_dir_real_paths{$needle_dir_path} = {
-                        real_path    => $needle_dir_real_path,
-                        real_path_id => $real_path_id,
-                    };
-                }
-                my $real_needle_info;
-                $real_needle_info
-                  = $schema->resultset('Needles')->find({dir_id => $real_needle_dir_id, filename => $filename})
-                  if $real_needle_dir_id;
-                $n = $real_needle_info if $real_needle_info;
-
-                $hash->{last_seen}  = $n->last_seen_time    || 'never';
-                $hash->{last_match} = $n->last_matched_time || 'never';
-
-                if (my $last_seen_module_id = $n->last_seen_module_id) {
-                    $hash->{last_seen_link} = $self->url_for(
-                        'admin_needle_module',
-                        module_id => $last_seen_module_id,
-                        needle_id => $n->id
-                    );
-                }
-                if (my $last_matched_module_id = $n->last_matched_module_id) {
-                    $hash->{last_match_link} = $self->url_for(
-                        'admin_needle_module',
-                        module_id => $last_matched_module_id,
-                        needle_id => $n->id
-                    );
-                }
-                push(@data, $hash);
-            }
-            return \@data;
+            my $needle_dirs = $self->schema->resultset('NeedleDirs');
+            my $needles     = $self->schema->resultset('Needles');
+            my %paths = {};
```

Comment:
maybe this hash should be written `my %path=()`

Reply:
Actually, just write it as `my %paths;` without any assignment. There's no such thing as an uninitialized hash in Perl; you always get an empty/default-initialized one.

Reply:
done